File: UIClient/src/main/java/cz/zelenikr/remotetouch/manager/ConnectionManager.java
package cz.zelenikr.remotetouch.manager;

import cz.zelenikr.remotetouch.Callback;
import cz.zelenikr.remotetouch.data.dto.event.CallEventContent;
import cz.zelenikr.remotetouch.data.dto.event.NotificationEventContent;
import cz.zelenikr.remotetouch.data.dto.event.SmsEventContent;
import cz.zelenikr.remotetouch.network.Client;
import cz.zelenikr.remotetouch.network.ConnectionStatus;
import cz.zelenikr.remotetouch.network.ContentRecivedListener;
import cz.zelenikr.remotetouch.network.SocketIOClient;
import cz.zelenikr.remotetouch.security.Hash;
import cz.zelenikr.remotetouch.security.SHAHash;
import org.jetbrains.annotations.NotNull;

import java.net.URI;
import java.net.URISyntaxException;
import java.util.HashSet;
import java.util.Set;

/**
 * Manages connection and communication with mobile device via {@link Client}.
 *
 * @author Roman Zelenik
 */
public final class ConnectionManager {
    private static final SettingsManager SETTINGS = SettingsManager.getInstance();
    private static final ConnectionManager INSTANCE = new ConnectionManager();

    public static ConnectionManager getInstance() {
        return INSTANCE;
    }

    private Client client;

    // ContentRecivedListener sets
    private Set<ContentRecivedListener<CallEventContent>> callReceivedListeners = new HashSet<>();
    private Set<ContentRecivedListener<NotificationEventContent>> notificationReceivedListeners = new HashSet<>();
    private Set<ContentRecivedListener<SmsEventContent>> smsReceivedListeners = new HashSet<>();

    // Connection state changed listeners
    private Set<Callback<ConnectionStatus>> connectionStateChangedListeners = new HashSet<>();

    /**
     * Connects to a device. Client is initialized just once when this method is called.
     * If client is already connected, nothing happen.
     */
    public void connect() {
        if (client == null) {
            client = initClient();
        }
        if (!client.isConnected()) {
            client.connect();
        }
    }

    /**
     * Disconnects a device.
     */
    public void disconnect() {
        if (client != null) {
            client.disconnect();
            client = null;
        }
    }

    /**
     * Disconnects and establishes new connection to a mobile device.
     */
    public void reconnect() {
        if (client != null) {
            disconnect();
        }
        connect();
    }

    public void registerConnectionStateChangedListener(@NotNull Callback<ConnectionStatus> listener) {
        connectionStateChangedListeners.add(listener);
    }

    public void registerCallReceivedListener(@NotNull ContentRecivedListener<CallEventContent> listener) {
        callReceivedListeners.add(listener);
    }

    public void registerNotificationReceivedListener(@NotNull ContentRecivedListener<NotificationEventContent> listener) {
        notificationReceivedListeners.add(listener);
    }

    public void registerSmsReceivedListener(@NotNull ContentRecivedListener<SmsEventContent> listener) {
        smsReceivedListeners.add(listener);
    }

    public void removeConnectionStateChangedListener(@NotNull Callback<ConnectionStatus> listener) {
        connectionStateChangedListeners.remove(listener);
    }

    public void removeCallReceivedListener(@NotNull ContentRecivedListener<CallEventContent> listener) {
        callReceivedListeners.remove(listener);
    }

    public void removeNotificationReceivedListener(@NotNull ContentRecivedListener<NotificationEventContent> listener) {
        notificationReceivedListeners.remove(listener);
    }

    public void removeSmsReceivedListener(@NotNull ContentRecivedListener<SmsEventContent> listener) {
        smsReceivedListeners.remove(listener);
    }

    private Client initClient() {
        SocketIOClient client = new SocketIOClient(loadClientToken(), loadServerAddress(), loadPairKey());
        client.setOnCallReceived(contents -> notifyCallReceived(contents));
        client.setOnNotificationReceived(contents -> notifyNotificationReceived(contents));
        client.setOnSMSReceived(contents -> notifySmsReceived(contents));
        client.setOnConnectionStatusChanged(status -> notifyConnectionStateChanged(status));
        return client;
    }

    private Void notifyConnectionStateChanged(ConnectionStatus status) {
        connectionStateChangedListeners.forEach(callback -> callback.call(status));
        return null;
    }

    private void notifyCallReceived(CallEventContent[] contents) {
        callReceivedListeners.forEach(contentRecivedListener -> contentRecivedListener.onReceived(contents));
    }

    private void notifyNotificationReceived(NotificationEventContent[] contents) {
        notificationReceivedListeners.forEach(contentRecivedListener -> contentRecivedListener.onReceived(contents));
    }

    private void notifySmsReceived(SmsEventContent[] contents) {
        smsReceivedListeners.forEach(contentRecivedListener -> contentRecivedListener.onReceived(contents));
    }

    private String loadPairKey() {
        return SETTINGS.getPairKey();
    }

    private URI loadServerAddress() {
        try {
            return SETTINGS.getServerAddress().toURI();
        } catch (URISyntaxException e) {
            //
            return null;
        }
    }

    private String loadClientToken() {
        return createToken(SETTINGS.getDeviceName(), loadPairKey());
    }

    private static String createToken(String device, String pairKey) {
        Hash hash = new SHAHash();
        return hash.hash(device + pairKey);
    }

    private ConnectionManager() {
    }
}

File: WebClient/src/manager/ConnectionManager.kt
package manager

import data.dto.event.NotificationEventContent
import lib.js.Callback
import log.Logger
import network.Client
import network.ConnectionStatus
import network.ContentReceivedListener
import network.SocketIOClient


/**
 * Manages connection and communication with mobile device via {@link Client}.
 *
 * @author Roman Zelenik
 */
object ConnectionManager {

    private val logger: Logger = Logger.getLogger("ConnectionManager")
    private var client: Client? = null

    // ContentReceivedListener sets
//    private val callReceivedListeners: Set<ContentReceivedListener<CallEventContent>> = setOf()
    private val notificationReceivedListeners: MutableSet<ContentReceivedListener<NotificationEventContent>> = mutableSetOf()
//    private val smsReceivedListeners: Set<ContentReceivedListener<SmsEventContent>> = setOf()

    // Connection state changed listeners
    private val connectionStateChangedListeners: MutableSet<Callback<ConnectionStatus>> = mutableSetOf()

    /**
     * Connects to a device. Client is initialized just once when this method is called.
     * If client is already connected, nothing happen.
     */
    fun connect() {
        if (client == null) {
            client = initClient();
        }
        if (!client!!.isConnected) {
            client!!.connect();
        }
    }

    /**
     * Disconnects a device.
     */
    fun disconnect() {
        client?.disconnect();
        client = null;
    }

    /**
     * Disconnects and establishes new connection to a mobile device.
     */
    fun reconnect() {
        if (client != null) {
            disconnect();
        }
        connect();
    }

    fun isConnected(): Boolean = client?.isConnected ?: false


    fun registerConnectionStateChangedListener(listener: Callback<ConnectionStatus>) {
        connectionStateChangedListeners += listener
    }

    /*
            fun registerCallReceivedListener(@NotNull ContentRecivedListener<CallEventContent> listener) {
                callReceivedListeners.add(listener);
            }
    */
    fun registerNotificationReceivedListener(listener: ContentReceivedListener<NotificationEventContent>) {
        notificationReceivedListeners += listener
    }

    /*
            fun registerSmsReceivedListener(@NotNull ContentRecivedListener<SmsEventContent> listener) {
                smsReceivedListeners.add(listener);
            }
*/
    fun removeConnectionStateChangedListener(listener: Callback<ConnectionStatus>) {
        connectionStateChangedListeners -= listener
    }

    /*
                fun removeCallReceivedListener(@NotNull ContentRecivedListener<CallEventContent> listener) {
                    callReceivedListeners.remove(listener);
                }
        */
    fun removeNotificationReceivedListener(listener: ContentReceivedListener<NotificationEventContent>) {
        notificationReceivedListeners -= listener
    }

    /*
            fun removeSmsReceivedListener(@NotNull ContentRecivedListener<SmsEventContent> listener) {
                smsReceivedListeners.remove(listener);
            }
        */
    private fun initClient(): Client {
        return SocketIOClient(loadClientToken(), loadServerAddress(), loadPairKey()).also {
            //        client.setOnCallReceived(contents -> notifyCallReceived(contents));
//            it.setOnNotificationReceived{notifyNotificationReceived(contents)}

//        client.setOnSMSReceived(contents -> notifySmsReceived(contents));
            it.setOnConnectionStatusChanged { status ->
                notifyConnectionStateChanged(status)
            }
        }
    }


    private fun notifyConnectionStateChanged(status: ConnectionStatus) {
//        logger.debug("notifyConnectionStateChanged: " + status.name)
        connectionStateChangedListeners.forEach { callback -> callback(status) }
    }

    /*
            private void notifyCallReceived(CallEventContent[] contents) {
                callReceivedListeners.forEach(contentRecivedListener -> contentRecivedListener.onReceived(contents));
            }
       */
    private fun notifyNotificationReceived(contents: Array<NotificationEventContent>) {
        notificationReceivedListeners.forEach { contentReceivedListener -> contentReceivedListener.onReceived(*contents) };
    }

//        private void notifySmsReceived(SmsEventContent[] contents) {
//            smsReceivedListeners.forEach(contentRecivedListener -> contentRecivedListener.onReceived(contents));
//        }

    private fun loadPairKey(): String {
        return "[B@6799e8e"
//        return SETTINGS.getPairKey();
    }

    private fun loadServerAddress(): String {
        return "localhost:8080/socket"
//            return SETTINGS.getServerAddress().toURI();
    }

    private fun loadClientToken(): String {
        return createToken("", loadPairKey())
//        return createToken(SETTINGS.getDeviceName(), loadPairKey());
    }

    private fun createToken(device: String, pairKey: String): String {
        return "R7eG00h/U9wVRkiITVOpZWQJT9E1YZmTAIXBuRbNfz4="
//        Hash hash = new SHAHash();
//        return hash.hash(device + pairKey);
    }

}
